src/v/features/migrators/cloud_storage_config.cc

```diff
@@ -45,10 +45,14 @@ ss::future<> cloud_storage_config::do_mutate() {
         auto cloud_storage_enabled
           = config::shard_local_cfg().cloud_storage_enabled();
 
-        auto remote_write_enabled
-          = model::is_archival_enabled(props.shadow_indexing.value_or(
-              model::shadow_indexing_mode::disabled))
-            || config::shard_local_cfg().cloud_storage_enable_remote_write();
+        bool remote_write_enabled{false};
+        if (props.shadow_indexing) {
+            remote_write_enabled = model::is_archival_enabled(
+              *(props.shadow_indexing));
+        } else {
+            remote_write_enabled
+              = config::shard_local_cfg().cloud_storage_enable_remote_write();
+        }
```

Comment:
Not sure I understand this change. It would be a potentially surprising change of behaviour on upgrade. If you had `remote.write=False` and `cloud_storage_enable_remote_write=True`, we'd stop archiving data after the upgrade.

Reply:
Hmm, wasn't that the intention of https://github.com/redpanda-data/redpanda/pull/6663, to change that behavior so that if the topic level config said false, we wouldn't upload even if the cluster config was true?

Reply:
It was. At that time I thought that the archival logic simply respects what's in the `ntp_config`, but that turned out not to be true. So https://github.com/redpanda-data/redpanda/pull/6663 fixed the bug in assigning to the `ntp_config`, but we'd need to merge https://github.com/redpanda-data/redpanda/pull/6713 to actually follow `ntp_config` at all times.

This was my bad. I didn't realise there were two separate issues at the time.